unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/NotificationMessagesHUD/ChatNotificationController.cs

```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+using System.Linq;
+using Cysharp.Threading.Tasks;
+using UnityEngine;
+using DCL.Interface;
+using DCL;
+
+public class ChatNotificationController : IHUD
+{
+    private DataStore dataStore;
+    private IChatController chatController;
+    private IMainChatNotificationsComponentView mainChatNotificationView;
+    private IUserProfileBridge userProfileBridge;
+    private BaseVariable<Transform> notificationPanelTransform => dataStore.HUDs.notificationPanelTransform;
+    private BaseVariable<HashSet<string>> visibleTaskbarPanels => dataStore.HUDs.visibleTaskbarPanels;
+    private const int FADEOUT_DELAY = 8000;
+    public CancellationTokenSource fadeOutCT = new CancellationTokenSource();
+
+    private UserProfile ownUserProfile;
+
+    public ChatNotificationController(DataStore dataStore, IMainChatNotificationsComponentView mainChatNotificationView, IChatController chatController, IUserProfileBridge userProfileBridge)
+    {
+        this.dataStore = dataStore;
+        this.chatController = chatController;
+        this.userProfileBridge = userProfileBridge;
+        this.mainChatNotificationView = mainChatNotificationView;
+        mainChatNotificationView.OnResetFade += ResetFadeOut;
+        ownUserProfile = userProfileBridge.GetOwn();
+        chatController.OnAddMessage += HandleMessageAdded;
+        notificationPanelTransform.Set(mainChatNotificationView.GetPanelTransform());
+        visibleTaskbarPanels.OnChange += VisiblePanelsChanged;
+    }
+
+    private void VisiblePanelsChanged(HashSet<string> newList, HashSet<string> oldList)
+    {
+        SetVisibility(newList.Count == 0);
+    }
+
+    private void HandleMessageAdded(ChatMessage message)
+    {
+        if (message.messageType != ChatMessage.Type.PRIVATE && message.messageType != ChatMessage.Type.PUBLIC) return;
+        if (message.sender == ownUserProfile.userId) return;
+
+        if (message.sender != null)
+        {
+            var profile = ExtractRecipient(message);
+            if (profile == null) return;
+            mainChatNotificationView.AddNewChatNotification(message, profile.userName, profile.face256SnapshotURL);
+        }
+        else
+        {
+            mainChatNotificationView.AddNewChatNotification(message);
+        }
+    }
+
+    public void ResetFadeOut(bool fadeOutAfterDelay = false)
+    {
+        mainChatNotificationView.ShowNotifications();
+        fadeOutCT.Cancel();
+        fadeOutCT = new CancellationTokenSource();
```

Comment:
A doubt: Is it necessary to create a new instance of the `CancellationTokenSource` after doing a `.Cancel()`? Is it because the cancellation sets the variable as `null` or something like that?

Reply:
It's because of how the CancellationTokenSource works, once cancelled that token is done, we need a new one